// ngx-timer/esm5/lib/countdown-timer/countdown-timer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { countDownTimerConfigModel, countDownTimerTexts } from './countdown-timer.model';
import { CountdownTimerService } from './countdown-timer.service';
var CountdownTimerComponent = /** @class */ (function () {
    function CountdownTimerComponent(countdownTimerService) {
        var _this = this;
        this.countdownTimerService = countdownTimerService;
        this.timerObj = {};
        //get timer value
        this.getTimerValue = (/**
         * @return {?}
         */
        function () {
            _this.timerSubscription = _this.countdownTimerService.getTimerValue().subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                _this.timerObj = Object.assign(res);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                console.log(error);
                console.log('Failed to get timer value');
            }));
        });
    }
    /**
     * @return {?}
     */
    CountdownTimerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getTimerValue();
        this.timerConfig = new countDownTimerConfigModel();
        this.timerTextConfig = new countDownTimerTexts();
        this.timerConfig = this.countDownTimerConfig ? Object.assign(this.countDownTimerConfig) : null;
        this.timerTextConfig = this.countDownTimerConfig && this.countDownTimerConfig.timerTexts ? Object.assign(this.countDownTimerConfig.timerTexts) : null;
    };
    /**
     * @return {?}
     */
    CountdownTimerComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.timerSubscription.unsubscribe();
    };
    CountdownTimerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'countdown-timer',
                    template: "<div [class]=\"timerConfig?.timerClass ? timerConfig.timerClass : ''\">\r\n  <span>{{timerObj.hours}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.hourText ? timerTextConfig.hourText : 'hh'}} </span> \r\n  <span>{{timerObj.mins}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.minuteText ? timerTextConfig.minuteText : 'mm'}}  </span> \r\n  <span>{{timerObj.seconds}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.secondsText ? timerTextConfig.secondsText : 'ss'}}</span>\r\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    CountdownTimerComponent.ctorParameters = function () { return [
        { type: CountdownTimerService }
    ]; };
    CountdownTimerComponent.propDecorators = {
        startTime: [{ type: Input }],
        countDownTimerConfig: [{ type: Input }]
    };
    return CountdownTimerComponent;
}());
export { CountdownTimerComponent };
if (false) {
    /** @type {?} */
    CountdownTimerComponent.prototype.startTime;
    /** @type {?} */
    CountdownTimerComponent.prototype.countDownTimerConfig;
    /**
     * @type {?}
     * @private
     */
    CountdownTimerComponent.prototype.timerSubscription;
    /** @type {?} */
    CountdownTimerComponent.prototype.timerObj;
    /** @type {?} */
    CountdownTimerComponent.prototype.timerConfig;
    /** @type {?} */
    CountdownTimerComponent.prototype.timerTextConfig;
    /** @type {?} */
    CountdownTimerComponent.prototype.getTimerValue;
    /**
     * @type {?}
     * @private
     */
    CountdownTimerComponent.prototype.countdownTimerService;
}
