// ngx-timer/esm5/lib/countdown-timer/countdown-timer.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var countDownTimerConfigModel = /** @class */ (function () {
    function countDownTimerConfigModel() {
    }
    return countDownTimerConfigModel;
}());
export { countDownTimerConfigModel };
if (false) {
    /** @type {?} */
    countDownTimerConfigModel.prototype.timerClass;
    /** @type {?} */
    countDownTimerConfigModel.prototype.timerTexts;
}
var countDownTimerTexts = /** @class */ (function () {
    function countDownTimerTexts() {
    }
    return countDownTimerTexts;
}());
export { countDownTimerTexts };
if (false) {
    /** @type {?} */
    countDownTimerTexts.prototype.hourText;
    /** @type {?} */
    countDownTimerTexts.prototype.minuteText;
    /** @type {?} */
    countDownTimerTexts.prototype.secondsText;
}
/** @enum {string} */
var TimerStaus = {
    START: "START",
    PAUSE: "PAUSE",
    STOP: "STOP",
};
export { TimerStaus };
