// ngx-timer/esm5/lib/countup-timer/countup-timer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { CountupTimerService } from './countup-timer.service';
import { countUpTimerConfigModel, timerTexts } from './countup-timer.model';
var CountupTimerComponent = /** @class */ (function () {
    function CountupTimerComponent(countupTimerService) {
        var _this = this;
        this.countupTimerService = countupTimerService;
        //Init
        this.timerObj = {};
        //get timer value
        this.getTimerValue = (/**
         * @return {?}
         */
        function () {
            _this.timerSubscription = _this.countupTimerService.getTimerValue().subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                _this.timerObj = Object.assign(res);
            }), (/**
             * @param {?} error
             * @return {?}
             */
            function (error) {
                console.log(error);
                console.log('Failed to get timer value');
            }));
        });
    }
    /**
     * @return {?}
     */
    CountupTimerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getTimerValue();
        this.timerConfig = new countUpTimerConfigModel();
        this.timerTextConfig = new timerTexts();
        this.timerConfig = this.countUpTimerConfig ? Object.assign(this.countUpTimerConfig) : null;
        this.timerTextConfig = this.countUpTimerConfig && this.countUpTimerConfig.timerTexts ? Object.assign(this.countUpTimerConfig.timerTexts) : null;
    };
    /**
     * @return {?}
     */
    CountupTimerComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.timerSubscription.unsubscribe();
    };
    CountupTimerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'countup-timer',
                    template: "<div [class]=\"timerConfig?.timerClass ? timerConfig.timerClass : ''\">\n  <span>{{timerObj.hours}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.hourText ? timerTextConfig.hourText : 'hh'}} </span> \n  <span>{{timerObj.mins}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.minuteText ? timerTextConfig.minuteText : 'mm'}}  </span> \n  <span>{{timerObj.seconds}}</span><span class=\"time-category\">{{timerTextConfig && timerTextConfig.secondsText ? timerTextConfig.secondsText : 'ss'}}</span>\n</div>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    CountupTimerComponent.ctorParameters = function () { return [
        { type: CountupTimerService }
    ]; };
    CountupTimerComponent.propDecorators = {
        startTime: [{ type: Input }],
        countUpTimerConfig: [{ type: Input }]
    };
    return CountupTimerComponent;
}());
export { CountupTimerComponent };
if (false) {
    /** @type {?} */
    CountupTimerComponent.prototype.startTime;
    /** @type {?} */
    CountupTimerComponent.prototype.countUpTimerConfig;
    /** @type {?} */
    CountupTimerComponent.prototype.timerObj;
    /**
     * @type {?}
     * @private
     */
    CountupTimerComponent.prototype.timerSubscription;
    /** @type {?} */
    CountupTimerComponent.prototype.timerConfig;
    /** @type {?} */
    CountupTimerComponent.prototype.timerTextConfig;
    /** @type {?} */
    CountupTimerComponent.prototype.getTimerValue;
    /**
     * @type {?}
     * @private
     */
    CountupTimerComponent.prototype.countupTimerService;
}
