// @covalent/markdown/esm2015/markdown-utils.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
/**
 * @param {?} str
 * @return {?}
 */
export function removeLeadingHash(str) {
    if (str) {
        return str.replace(/^#+/, '');
    }
    return undefined;
}
/**
 * @param {?} str
 * @return {?}
 */
export function removeTrailingHash(str) {
    if (str) {
        return str.replace(/\#.*/, '');
    }
    return undefined;
}
/**
 * @param {?} str
 * @return {?}
 */
export function normalizeAnchor(str) {
    if (str) {
        return removeLeadingHash(str.replace(/(_|-|\s)+/g, '')).toLowerCase();
    }
    return undefined;
}
/**
 * @param {?} element
 * @param {?} anchor
 * @return {?}
 */
export function scrollToAnchor(element, anchor) {
    if (element && anchor) {
        /** @type {?} */
        const normalizedAnchor = normalizeAnchor(anchor);
        /** @type {?} */
        const parent = element.parentElement;
        /** @type {?} */
        let headingToJumpTo;
        /** @type {?} */
        const headingWithinComponent = element.querySelector(`[id="${normalizedAnchor}"]`);
        if (headingWithinComponent) {
            headingToJumpTo = headingWithinComponent;
        }
        else if (parent) {
            headingToJumpTo = parent.querySelector(`[id="${normalizedAnchor}"]`);
        }
        if (headingToJumpTo) {
            headingToJumpTo.scrollIntoView({ behavior: 'auto' });
        }
        else {
            // TODO: leave this warning?
            // tslint:disable-next-line
            console.warn(`Could not jump to heading '${anchor}'`);
        }
    }
}
/**
 * @param {?} anchor
 * @return {?}
 */
export function isAnchorLink(anchor) {
    if (anchor) {
        return anchor.getAttribute('href').startsWith('#');
    }
}
/**
 * @param {?} githubHref
 * @return {?}
 */
export function rawGithubHref(githubHref) {
    if (githubHref) {
        /** @type {?} */
        const url = new URL(githubHref);
        if (url.pathname.startsWith('/raw/')) {
            return githubHref;
        }
        else {
            url.hostname = 'raw.githubusercontent.com';
            url.pathname = url.pathname.split('/blob', 2).join('');
            return url.href;
        }
    }
    return undefined;
}
/**
 * @param {?} href
 * @return {?}
 */
export function isGithubHref(href) {
    try {
        /** @type {?} */
        const temp = new URL(href);
        return temp.hostname === 'github.com';
    }
    catch (_a) {
        return false;
    }
}
