// ngx-timer/esm5/lib/ngx-timer.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CountupTimerComponent } from './countup-timer/countup-timer.component';
import { CommonModule } from '@angular/common';
import { CountdownTimerComponent } from './countdown-timer/countdown-timer.component';
import { CountupTimerService } from './countup-timer/countup-timer.service';
import { CountdownTimerService } from './countdown-timer/countdown-timer.service';
import { StopWatchComponent } from './stop-watch/stop-watch.component';
import { StopWatchService } from './stop-watch/stop-watch.service';
var NgxTimerModule = /** @class */ (function () {
    function NgxTimerModule() {
    }
    NgxTimerModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CountupTimerComponent,
                        CountdownTimerComponent,
                        StopWatchComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [
                        CountupTimerComponent,
                        CountdownTimerComponent,
                        StopWatchComponent
                    ],
                    entryComponents: [
                        CountupTimerComponent,
                        CountdownTimerComponent,
                        StopWatchComponent
                    ],
                    providers: [
                        CountupTimerService,
                        CountdownTimerService,
                        StopWatchService
                    ]
                },] }
    ];
    return NgxTimerModule;
}());
export { NgxTimerModule };
